cou/utils/juju_utils.py

```diff
@@ -396,3 +397,101 @@ async def async_upgrade_charm(
         revision=revision,
         switch=switch,
     )
+
+
+# pylint: disable=too-few-public-methods
+class JujuWaiter:
+    """Enhanced version of wait_for_idle.
+
+    Usage:
+        JujuWaiter(model).wait(120)
+    """
+
+    DEFAULT_TIMEOUT: int = 3600
+    MODEL_IDLE_PERIOD: int = 30
+    JUJU_IDLE_CHECK_PERIOD: int = 40
+
+    class TimeoutException(Exception):
+        """Own timeout exception."""
+
+    def __init__(self, model: Model):
+        """Initialize.
+
+        :param model: model to wait
+        :type model: Model
+        """
+        self.model = model
+        self.model_name = self.model.info.name
+        self.timeout = timedelta(seconds=JujuWaiter.DEFAULT_TIMEOUT)
+        self.start_time = datetime.now()
+        self.log = logging.getLogger(self.__class__.__name__)
+
+    async def wait(self, timeout_seconds: Optional[int] = None) -> None:
+        """Wait model to stabilize.
+
+        :param timeout_seconds: wait model till timeout_seconds. If passed raise TimeoutException
+        :type timeout_seconds: Optional[int]
+        :return: None
+        :raises TimeoutException: if timeout_seconds passed.
+        :raises JujuMachineError: if it has a machine exception
+        :raises JujuAgentError: if it has an agent exception
+        :raises JujuUnitError: if it has unit exception
+        :raises JujuAppError: if it has application exception
+        """
+        self.start_time = datetime.now()
+        if timeout_seconds:
+            self.timeout = timedelta(seconds=timeout_seconds)
+        self.log.debug("Waiting to stabilize in %s seconds", self.timeout)
+
+        while True:
+            await self._ensure_model_connected()
+            try:
+                await self.model.wait_for_idle(
+                    idle_period=JujuWaiter.MODEL_IDLE_PERIOD,
+                    timeout=JujuWaiter.JUJU_IDLE_CHECK_PERIOD,
```

Comment:
Is this correct? Maybe this
```python
check_freq=JujuWaiter.JUJU_IDLE_CHECK_PERIOD, timeout=self.timeout
```

Reply:
It is correct we want it to be count as idle if it is idle more than `MODEL_IDLE_PERIOD` and we want it to timeout after `JUJU_IDLE_CHECK_PERIOD`

Reply:
Can we provided some description? Just to describe why we are using check period for timeout. Thanks

Reply:
Added extra comments for the constants.